eta/detectors/tfmodels_segmenters.py

```diff
@@ -0,0 +1,197 @@
+    omitted fields present in the default deployment config for the published
+    model will be automatically populated.
+
+    Attributes:
+        model_name: the name of the published model to load. If this value is
+            provided, `model_path` does not need to be
+        model_path: the path to a TF SavedModel in `.pb` format to load. If
+            this value is provided, `model_name` does not need to be
+        labels_path: the path to the labels map for the model
+        confidence_thresh: a confidence threshold to apply to candidate
+            detections
+        mask_thresh: the threshold to use when generating the instance masks
+            for detections
+    '''
+
+    def __init__(self, d):
+        self.model_name = self.parse_string(d, "model_name", default=None)
+        self.model_path = self.parse_string(d, "model_path", default=None)
+
+        # Loads any default deployment parameters, if possible
+        if self.model_name:
+            d = self.load_default_deployment_params(d, self.model_name)
+
+        self.labels_path = etau.fill_config_patterns(
+            self.parse_string(d, "labels_path"))
+        self.confidence_thresh = self.parse_number(
+            d, "confidence_thresh", default=None)
+        self.mask_thresh = self.parse_number(d, "mask_thresh", default=0.5)
+
+        self._validate()
+
+    def _validate(self):
+        if not self.model_name and not self.model_path:
+            raise ConfigError(
+                "Either `model_name` or `model_path` must be provided")
+
+
+class TFModelsSegmenter(ObjectDetector, UsesTFSession):
+    '''Interface to the instance segmentation models from the TF-Models
+    detection library at
+    https://github.com/tensorflow/models/tree/master/research/object_detection.
+
+    This class uses `eta.core.tfutils.UsesTFSession` to create TF sessions, so
+    it automatically applies settings in your `eta.config.tf_config`.
+
+    Instances of this class must either use the context manager interface or
+    manually call `close()` when finished to release memory.
+    '''
+
+    def __init__(self, config):
+        '''Creates a TFModelsSegmenter instance.
+
+        Args:
+            config: a TFModelsSegmenterConfig instance
+        '''
+        self.config = config
+        UsesTFSession.__init__(self)
+
+        if self.config.model_name:
+            # Downloads the published model, if necessary
+            model_path = etam.download_model(self.config.model_name)
+        else:
+            model_path = self.config.model_path
+
+        # Load model
+        self._tf_graph = self._build_graph(model_path)
+        self._sess = self.make_tf_session(graph=self._tf_graph)
+
+        # Load labels
+        label_map = gool.load_labelmap(self.config.labels_path)
+        categories = gool.convert_label_map_to_categories(
+            label_map, max_num_classes=90, use_display_name=True)
```

Comment:
Where does this `90` come from?

Reply:
good catch. For reasons I don't understand, this method forces one to provide an upper bound on the number of classes to load from the label map, and 90 is the number from the COCO label map. I discovered that setting this to infinity works, so I did that.

90 = magic number. inf = cool.